src/python/pants/engine/internals/native_engine.pyi

```diff
@@ -248,6 +250,97 @@ class PyExecutor:
     def to_borrowed(self) -> PyExecutor: ...
     def shutdown(self, duration_secs: float) -> None: ...
 
+# ------------------------------------------------------------------------------
+# Target
+# ------------------------------------------------------------------------------
+
+# Type alias to express the intent that the type should be immutable and hashable. There's nothing
+# to actually enforce this, outside of convention.
+ImmutableValue = Any
```

Comment:
FYI The Python std library (and therefore core devs) have roughly said that hashability implies immutability. Search https://docs.python.org/3/library/dataclasses.html for "mutability".

So in that regard, this could be an alias for `typing.Hashable`

Reply:
That's potentially backwards incompatible... but assuming that `mypy` doesn't misbehave, it should only catch bugs?

Reply:
I didn't look into it, but there were at least 10.

My goal with these port PRs is not to change behavior (unless it makes it _easier_ to accomplish them), since my feeling right now is that getting out of our performance hole is the most important goal.